v-next/hardhat/src/internal/cli/main.ts

```diff
@@ -265,3 +269,192 @@ function getTaskFromCliArguments(
 
   return task;
 }
+
+function parseTaskArguments(
+  cliArguments: string[],
+  usedCliArguments: boolean[],
+  task: Task,
+): Record<string, any> {
+  const taskArguments: Record<string, unknown> = {};
+
+  parseNamedParameters(cliArguments, usedCliArguments, task, taskArguments);
+
+  parsePositionalAndVariadicParameters(
+    cliArguments,
+    usedCliArguments,
+    task,
+    taskArguments,
+  );
+
+  usedCliArguments.forEach((arg, index) => {
+    if (!arg) {
+      throw new HardhatError(HardhatError.ERRORS.ARGUMENTS.UNUSED_ARGUMENT, {
+        value: cliArguments[index],
+      });
+    }
+  });
+
+  return taskArguments;
+}
+
+function parseNamedParameters(
+  cliArguments: string[],
+  usedCliArguments: boolean[],
+  task: Task,
+  taskArguments: Record<string, any>,
+) {
+  for (let i = 0; i < cliArguments.length; i++) {
+    if (usedCliArguments[i]) {
+      continue;
+    }
+
+    if (cliArguments[i] === "--") {
+      // A standalone '--' is ok because it is used to separate CLI tool arguments from script arguments,
+      // ensuring the tool passes subsequent options directly to the script.
+      continue;
```

Comment:
Shouldn't this be a `break`? 

I'll try to validate it with a test.

Reply:
Yes, it should be a break. For example, this parsers correctly, when it should be an error:

```ts
const hre = await createHardhatRuntimeEnvironment({
      tasks: [
        task(["task"])
          .addNamedParameter({ name: "param" })
          .addNamedParameter({
            name: "flag",
            type: ParameterType.BOOLEAN,
          })
          .addNamedParameter({
            name: "optionFlag",
          })
          .setAction(() => {})
          .build(),
      ],
    });

    const command = "npx hardhat task -- --param asd";
    const cliArguments = command.split(" ").slice(2);
    const unusedCliArguments = new Array(cliArguments.length).fill(false);
```

It prints:
```
{
  task: ResolvedTask {
    id: [ 'task' ],
    description: '',
    actions: [ [Object] ],
    namedParameters: Map(3) {
      'param' => [Object],
      'flag' => [Object],
      'optionFlag' => [Object]
    },
    positionalParameters: [],
    pluginId: undefined,
    subtasks: Map(0) {}
  },
  taskArguments: { param: 'asd' }
}
```

But everything after `--` should be considered positional params, and fail, as this task doesn't have any.

Reply:
[Done in this commit](https://github.com/NomicFoundation/hardhat/pull/5300/commits/49e9e2c630dd244de22da5ad5ec843827ee4769e)